text/0000-make-cargo-install-extensible.md

```diff
@@ -0,0 +1,122 @@
+
+Additionally, this proposal will make `cargo install` more useful for the larger set of rust applications that leverage the larger, often preferred, platform-specific distribution workflows, which typically want to install into a temporary directory and package the result. This point is discussed further in the “Rationale and Alternatives” section.
+
+Fundamentally, the aim of this RFC is aligned directly with Rust’s value of “productivity”. It aims to make both building and using small developer CLI tools simpler and easier which has a direct positive effect on the ability of all to improve their workflow. 
+
+## Motivation
+
+The motivation for this RFC is the desire to improve developer experience for those writing and using applications built in the [2018 Rust Roadmap](https://blog.rust-lang.org/2018/03/12/roadmap.html) identified [domains](https://blog.rust-lang.org/2018/03/12/roadmap.html#four-target-domains), specifically command line applications. The idea for this RFC originated in a discussion amongst the [CLI Working Group](https://internals.rust-lang.org/t/announcing-the-cli-working-group/6872) about [CLI application distribution](https://paper.dropbox.com/doc/CLI-WG-Berlin-2v99dJ7g6QVkGoT5VavPY) at the Rust All Hands in Berlin.
+
+We were discussing packaging of applications and it became clear that this fundamentally depends on the installation process setting up and generating additional files. Currently applications typically mention such steps in their documentation, leaving them to the user to manually run. To streamline this process for both tool creators and users we wanted to provide reusable functionality to automate those steps instead of providing yet more documentation to follow.
+
+## Guide-level explanation
+
+When running `cargo install`, in addition to installing the compiled binaries, cargo will additionally build and run an `install.rs` script, if one exists. This script exists primarily to install additional files; it may also generate those additional files, such as running code to generate a [man](https://en.wikipedia.org/wiki/Man_page) [](https://en.wikipedia.org/wiki/Man_page)[page](https://en.wikipedia.org/wiki/Man_page) or [shell completion file](https://en.wikipedia.org/wiki/Command-line_completion).
+
+In the common case, crates should make use of common infrastructure crates to implement this extended installation functionality; for instance, a common crate could provide the functionality of generating shell completions from a command-line argument parser. To support this case, we should support and strongly encourage the use of a `metainstall` mechanism that allows one of a crate’s dependencies to provide the `install.rs` functionality, by analogy with [metabuild](https://github.com/rust-lang/rfcs/pull/2196/).
+
+### New Named Concepts
+
+- `install.rs`: a file that contains a set of instructions to occur after `cargo install` is run.
```

Comment:
To @est31's point on uninstallation, should `uninstall.rs` be required if there is a `install.rs`?
Of course, a crate can cheat and have `uninstall.rs` which is not the inverse of `install.rs`, but it is better than nothing?

Reply:
Yeah we discussed this issue at length. We're not sure if requiring the uninstall.rs is better or worse as I would think, personally that it gives the appearance of a guarantee that we cannot make, i.e. that it's not the inverse of `install.rs` by I'm open to being convinced otherwise.

Reply:
were any thoughts made for making/mandating that install.rs be reversible/isomorphic by design to allow the undoing automatically?

Reply:
I agree with @ashleygwilliams here: mandating (or even supporting) `uninstall.rs` is no guarantee for `uninstall.rs` actually uninstalling all of the artifacts. We should go either with the minicargo or monolithic approaches IMO. As for isomorphism @fbstj do you have any precedent of an isomorphic scripting language or similar that you can share? Also, we would have to make sure that all of the information that enters the program will stay the same, otherwise even if we have an isomorphic subset of Rust, we won't be able to guarantee equal outcome.

Reply:
@fbstj Is that actually possible tho? Unless we have some snapshot of how things were before, `install.rs` being possibly non-deterministic should not allow you to find an inverse function / program. Of course you could require that the `main` of an `install.rs` file be a `const fn` which would get you determinism, but even then - we'd have to rule out many good programs to make reversibility efficient or decidable, no? And you'd need to extend the type system to rule out such good programs.

@ashleygwilliams Given that `install.rs` and indeed `build.rs` could do all sorts of things to your computer, including invoking UB, I think that guarantees were never something that we could make on this front, and that it's all inherently trust based. I think that if we clearly communicate that this is all best-effort, then it is better to require `uninstall.rs` scripts and build great helper crates around this.

Reply:
> guarantees were never something that we could make on this front, and that it's all inherently trust based.

Right now we still have the option to sandbox `install.rs` and friends to one directory where they can put stuff in, asking cargo to do the actual installation for them so that uninstallation is possible.

But even if we decide to not sandbox, this argument sounds too much like: "you need to already trust them so making it easier for them to not introduce bugs is not important".